libraries/AC_Sprayer/AC_Sprayer.cpp

```diff
@@ -66,34 +66,43 @@ AC_Sprayer::AC_Sprayer(const AP_InertialNav* inav) :
     // To-Do: ensure that the pump and spinner servo channels are enabled
 }
 
-void AC_Sprayer::enable(bool true_false)
+void AC_Sprayer::run(const bool true_false)
 {
     // return immediately if no change
-    if (true_false == _enabled) {
+    if (true_false == _flags.running) {
         return;
     }
 
-    // set enabled/disabled parameter (in memory only)
-    _enabled = true_false;
+    // set flag indicate whether spraying is permitted:
+    // do not allow running to be set to true if we are currently not enabled
+    _flags.running = true_false && _enabled;
 
     // turn off the pump and spinner servos if necessary
-    if (!_enabled) {
-        // send output to pump channel
-        // To-Do: change 0 below to radio_min of pump servo
-        RC_Channel_aux::set_radio_to_min(RC_Channel_aux::k_sprayer_pump);
-
-        // send output to spinner channel
-        // To-Do: change 0 below to radio_min of spinner servo
-        RC_Channel_aux::set_radio_to_min(RC_Channel_aux::k_sprayer_spinner);
+    if (!_flags.running) {
+        stop_spraying();
     }
 }
 
+void AC_Sprayer::stop_spraying()
+{
+    // send output to pump channel
+    RC_Channel_aux::set_radio_to_min(RC_Channel_aux::k_sprayer_pump);
+
+    // send output to spinner channel
+    RC_Channel_aux::set_radio_to_min(RC_Channel_aux::k_sprayer_spinner);
+
+    _flags.spraying = false;
+}
+
 /// update - adjust pwm of servo controlling pump speed according to the desired quantity and our horizontal speed
 void
 AC_Sprayer::update()
 {
-    // exit immediately if we are disabled (perhaps set pwm values back to defaults)
-    if (!_enabled) {
+    // exit immediately if we are disabled or shouldn't be running
+    if (!_enabled || !running()) {
+        if (running()) {
+            run(false);
```

Comment:
The if can be removed, the *run* correctly checks if it is already not running and does nothing in that case.